vms/platformvm/txs/mempool/mempool.go

```diff
@@ -4,42 +4,326 @@
+	}
+
+	m := &Mempool{
+		weights:     weights,
+		gasCapacity: gasCapacity,
+		avaxAssetID: avaxAssetID,
+		maxHeap: heap.NewMap[ids.ID, heapTx](func(a, b heapTx) bool {
+			return a.GasPrice > b.GasPrice
+		}),
+		minHeap: heap.NewMap[ids.ID, heapTx](func(a, b heapTx) bool {
+			return a.GasPrice < b.GasPrice
+		}),
+		consumedUTXOs:      setmap.New[ids.ID, ids.ID](),
+		droppedTxIDs:       lru.NewCache[ids.ID, error](64),
+		numTxsMetric:       numTxsMetric,
+		gasAvailableMetric: gasAvailableMetric,
 	}
-	pool := txmempool.New[*txs.Tx](
-		metrics,
-	)
-	return &Mempool{Mempool: pool}, nil
+
+	m.cond = lock.NewCond(&m.lock)
+	return m, nil
 }
 
 func (m *Mempool) Add(tx *txs.Tx) error {
-	switch tx.Unsigned.(type) {
-	case *txs.AdvanceTimeTx:
-		return ErrCantIssueAdvanceTimeTx
-	case *txs.RewardValidatorTx:
-		return ErrCantIssueRewardValidatorTx
-	default:
-		return m.Mempool.Add(tx)
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.consumedUTXOs.HasOverlap(tx.InputIDs()) {
+		return mempool.ErrConflictsWithOtherTx
+	}
+
+	ins, outs, producedAVAX, err := utxo.GetInputOutputs(tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to get utxos: %w", err)
+	}
+
+	consumedAVAX := uint64(0)
+	for _, utxo := range ins {
+		if utxo.AssetID() != m.avaxAssetID {
+			continue
+		}
+
+		// The caller should verify txs but perform overflow checks anyway
+		consumedAVAX, err = math.Add(consumedAVAX, utxo.In.Amount())
+		if err != nil {
+			return fmt.Errorf("failed to add consumed AVAX: %w", err)
+		}
+	}
+
+	if consumedAVAX == 0 {
+		return errMissingConsumedAVAX
+	}
+
+	for _, utxo := range outs {
+		if utxo.AssetID() != m.avaxAssetID {
+			continue
+		}
+
+		// The caller should verify txs but perform overflow checks anyway
+		producedAVAX, err = math.Add64(producedAVAX, utxo.Out.Amount())
+		if err != nil {
+			return fmt.Errorf("failed to add produced AVAX: %w", err)
+		}
+	}
+
+	gasUsed, err := m.meter(tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to meter tx: %w", err)
+	}
+
+	heapTx := heapTx{
+		Tx:       tx,
+		gasUsed:  gasUsed,
+		GasPrice: float64(consumedAVAX-producedAVAX) / float64(gasUsed),
+	}
+
+	// Try to evict lower gas priced txs if we do not have enough remaining gas
+	// capacity
+	if next := m.currentGas + gasUsed; next > m.gasCapacity {
+		gasToFree := next - m.gasCapacity
+		if err := m.tryEvictTxs(gasToFree, heapTx); err != nil {
+			return err
+		}
+	}
+
+	m.maxHeap.Push(tx.TxID, heapTx)
+	m.minHeap.Push(tx.TxID, heapTx)
+	m.consumedUTXOs.Put(tx.TxID, tx.InputIDs())
+	m.droppedTxIDs.Evict(tx.TxID)
+	m.currentGas += gasUsed
+
+	m.updateMetrics()
+	m.cond.Broadcast()
+
+	return nil
+}
+
+func (m *Mempool) tryEvictTxs(gasToFree gas.Gas, txToAdd heapTx) error {
+	gasFreed := gas.Gas(0)
+	toEvict := make([]ids.ID, 0)
+
+	for _, tx := range m.minHeap.Iterator() {
+		// Try to evict lower priced txs to make room for the new tx
+		if tx.GasPrice >= txToAdd.GasPrice {
+			continue
```

Comment:
The next `tx` is guaranteed to have a `>=` `GasPrice` as well because we are iterating in sorted order.
```suggestion
			break
```